### DISCUSSION TITLE
Visualising data in different orientations?

### DISCUSSION THREAD

**Discussion by pbartholomew08** [ornladios/ADIOS2#3900] [Q&A] [answered]
Hi, we've added ADIOS2 as an I/O backend for our pencil-based code and it is working well. Recently we discovered some unexpected (to us) behaviour when visualising the output and wondered if this was correct or not.

Briefly, in the pencil-based decomposition the 3-D ``(nx*ny*nz)`` domain is decomposed along an axis at a time, for example in the X orientation each rank would have access to data ``(1:nx, j1X_rank:jnX_rank, k1X_rank:knX_rank)``, similarly in Y ``(i1Y_rank:inY_rank, 1:ny, k1Y_rank:knX_rank)`` and analogously for Z.

We support I/O from any of the 3 orientations and if we open a separate ``.bp4`` directory for each one then we see the visualisation expetcted (using rank ID as data)
![Screenshot 2023-11-07 at 10 25 53](https://github.com/ornladios/ADIOS2/assets/7126672/fc7e2394-04cc-41bc-a3a4-2367ab49cada)
![Screenshot 2023-11-07 at 10 25 49](https://github.com/ornladios/ADIOS2/assets/7126672/1701e28f-fe82-4371-be77-fda0f3cd1d73)
![Screenshot 2023-11-07 at 10 25 45](https://github.com/ornladios/ADIOS2/assets/7126672/918320d3-4b07-40f0-b4de-877c69c2a7a6)

However, if we use a single ``.bp4`` directory to output the same data then the output remains oriented in the first orientation when visualised. Printing the contents of the ``.bp4`` directory with ``bpls`` it looks like the data is written correctly for each orientation and we can write and read the data back correctly within our code.

Looking at the help for ``bpls`` I see there is a ``mesh`` option could our visualisation issue related to the ``.bp4`` mesh when written as a single output, and is this the expected behaviour?

**Comment by caitlinross** [comment]
It sounds like your bp file may be fine, but perhaps the reader in ParaView isn't handling things correctly. Could you share the single bp file that has all of the data where ParaView will only visualize one orientation? I think that will help me understand better what's going on.

**Reply by pbartholomew08** [reply]
Apologies for the delay, please find the file attached. When I open this in paraview, each of u1.dat, u2.dat, u3.dat looks like u1.dat (last image in my above post).
[out.bp4.zip](https://github.com/ornladios/ADIOS2/files/13301422/out.bp4.zip)

**Reply by caitlinross** [reply]
Thanks! I'm at a conference all next week, so I may not get a chance to really dig into what's going on with this until the week of the 20th.

**Reply by pbartholomew08** [reply]
Not to worry, it is really an edge case that we came across in our testing and just weren't sure if we had an error in our code. I appreciate the help

**Reply by pbartholomew08** [reply]
p.s. if it helps, I'm using Paraview 5.9.1 on a Mac

**Reply by caitlinross** [reply, marked as answer]
I don't have the problem fixed yet, but confirmed that I'm seeing the same thing. I'm certain it's purely an issue with the reader in ParaView and not an issue with how you're writing the file. When I load the file, I tested just loading a single array and setting the Image Dimension property to that same array and in that case the visualization is correct. But loading more than one array, it appears to go with the dimensions of the first array in the list, even if I manually change the Image Dimensions property, so I think something isn't respecting the Image Dimension property. I'll work on fixing it, though FYI the change would go into 5.12 (if I can get the fix before they finish the release candidate stage I think they'll let me add it to 5.12, otherwise it would be 5.12.1). I don't think we continue to make patch releases on old versions. 

At least there is a workaround for now, only load the array you're interested in and make sure both the Image Dimension and Time step array properties are also set to the correct array name. 

![Screenshot 2023-11-27 at 5 21 33 PM](https://github.com/ornladios/ADIOS2/assets/12058301/786c1b5b-6230-4f4a-b4c6-2659db2c65ab)

**Reply by pbartholomew08** [reply]
Thanks very much for investigating, and for finding the work around (I hadn't thought to try that). Upgrading paraview isn't a problem, I just haven't been keeping up with the new releases.

**Reply by caitlinross** [reply]
While debugging this, I ended up finding some other issues with this reader. All of the fixes are in this [VTK MR](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/10749). So once this is in (and makes it to ParaView), you'd be able to load all arrays and then you'd only have to change the Image Dimension property when you want to change which orientation you want to use.

**Reply by pbartholomew08** [reply]
Awesome, thank you!
